File: flutter_android/lib/simulation/auto_simulator.dart
import 'dart:math';

import '../game/case_model.dart';
import '../game/game_config.dart';
import '../game/game_session.dart';
import 'simulation_result.dart';

enum AutoStrategy {
  conservative,
}

class AutoSimulator {
  AutoSimulator({
    Random? random,
  }) : _random = random ?? Random();

  final Random _random;

  SimulationResult run({
    required int games,
    AutoStrategy strategy = AutoStrategy.conservative,
  }) {
    var totalHostProfit = 0;
    var totalPlayerAward = 0;
    var dealCount = 0;
    var finalChoiceCount = 0;

    for (var i = 0; i < games; i++) {
      final session = GameSession(random: _random);
      final playerPick = _random.nextInt(GameConfig.caseCount) + 1;
      session.choosePlayerCase(playerPick);

      while (session.phase == GamePhase.openingCases) {
        final candidates = session.cases
            .where((gameCase) => gameCase.status == CaseStatus.closed)
            .toList();
        final pick = candidates[_random.nextInt(candidates.length)];
        session.openCase(pick.number);

        if (session.phase == GamePhase.bankerOffer &&
            _shouldAcceptOffer(session)) {
          final result = session.acceptOffer();
          totalHostProfit += result.hostProfit;
          totalPlayerAward += result.awardedAmount;
          dealCount++;
          break;
        }

        if (session.phase == GamePhase.bankerOffer) {
          session.rejectOffer();
        }
      }

      if (session.phase == GamePhase.finalChoice) {
        final result = _random.nextBool()
            ? session.keepPlayerCase()
            : session.swapPlayerCase();
        totalHostProfit += result.hostProfit;
        totalPlayerAward += result.awardedAmount;
        finalChoiceCount++;
      }
    }

    return SimulationResult(
      games: games,
      totalHostProfit: totalHostProfit,
      totalPlayerAward: totalPlayerAward,
      dealCount: dealCount,
      finalChoiceCount: finalChoiceCount,
    );
  }

  bool _shouldAcceptOffer(GameSession session) {
    final playerCase = session.cases.firstWhere(
      (gameCase) => gameCase.number == session.playerCaseNumber,
    );
    final remaining = session.remainingAmounts;
    final expectedValue =
        remaining.fold<int>(0, (sum, amount) => sum + amount) / remaining.length;

    return session.currentOffer >= expectedValue * 0.9 ||
        session.currentOffer >= playerCase.amount * 1.5;
  }
}

File: simulation/auto_simulator.py
import random

from game import CaseStatus, GameConfig, GamePhase, GameSession

from .models import SimulationResult
from .player_strategy import (
    DEFAULT_PLAYER_MIX,
    PlayerMix,
    PlayerType,
    choose_player_type,
    should_accept_offer,
)


class AutoSimulator:
    def __init__(self, rng: random.Random | None = None) -> None:
        self._rng = rng or random.Random()

    def run(
        self,
        games: int,
        player_mix: tuple[PlayerMix, ...] = DEFAULT_PLAYER_MIX,
    ) -> SimulationResult:
        config = GameConfig()
        total_host_profit = 0
        total_player_award = 0
        deal_count = 0
        final_choice_count = 0
        player_type_counts = {player_type: 0 for player_type in PlayerType}
        player_type_host_profit = {player_type: 0 for player_type in PlayerType}

        for _ in range(games):
            player_type = choose_player_type(self._rng, player_mix)
            player_type_counts[player_type] += 1
            session = GameSession(config=config, rng=self._rng)
            player_pick = self._rng.randint(1, config.case_count)
            session.choose_player_case(player_pick)

            while session.phase == GamePhase.OPENING_CASES:
                candidates = [
                    case
                    for case in session.cases
                    if case.status == CaseStatus.CLOSED
                ]
                pick = self._rng.choice(candidates)
                session.open_case(pick.number)

                if session.phase == GamePhase.BANKER_OFFER:
                    if should_accept_offer(session, player_type, config):
                        result = session.accept_offer()
                        total_host_profit += result.host_profit
                        total_player_award += result.awarded_amount
                        player_type_host_profit[player_type] += result.host_profit
                        deal_count += 1
                        break
                    session.reject_offer()

            if session.phase == GamePhase.FINAL_CHOICE:
                result = (
                    session.keep_player_case()
                    if self._rng.choice((True, False))
                    else session.swap_player_case()
                )
                total_host_profit += result.host_profit
                total_player_award += result.awarded_amount
                player_type_host_profit[player_type] += result.host_profit
                final_choice_count += 1

        return SimulationResult(
            games=games,
            total_host_profit=total_host_profit,
            total_player_award=total_player_award,
            deal_count=deal_count,
            final_choice_count=final_choice_count,
            player_type_counts=player_type_counts,
            player_type_host_profit=player_type_host_profit,
        )
